Objects/typeobject.c

```diff
@@ -3398,15 +3398,64 @@ get_bases_tuple(PyObject *bases_in, PyType_Spec *spec)
     return PyTuple_Pack(1, bases_in);
 }
 
+static inline int
+check_basicsize_includes_size_and_offsets(PyTypeObject* type) {
+    if (type->tp_alloc != PyType_GenericAlloc) {
+        // Custom allocators can ignore tp_basicsize
+        return 1;
+    }
+
+    if (type->tp_base && type->tp_base->tp_basicsize > type->tp_basicsize) {
+        PyErr_Format(PyExc_TypeError,
+                    "tp_basicsize for type '%s' (%d) is too small for base '%s' (%d)",
+                    type->tp_name, type->tp_basicsize,
+                    type->tp_base->tp_name, type->tp_base->tp_basicsize);
+        return 0;
+    }
+    if (type->tp_weaklistoffset + (Py_ssize_t)sizeof(PyObject*) > (Py_ssize_t)type->tp_basicsize) {
+        PyErr_Format(PyExc_TypeError,
+                    "tp_basicsize for type '%s' (%d) is too small for weaklist offset %d",
+                    type->tp_name, type->tp_basicsize, type->tp_weaklistoffset);
+        return 0;
+    }
+    if (type->tp_dictoffset + (Py_ssize_t)sizeof(PyObject*) > (Py_ssize_t)type->tp_basicsize) {
+        PyErr_Format(PyExc_TypeError,
+                    "tp_basicsize for type '%s' (%d) is too small for dict offset %d",
+                    type->tp_name, type->tp_basicsize, type->tp_dictoffset);
+        return 0;
+    }
+    if (type->tp_vectorcall_offset + (Py_ssize_t)sizeof(vectorcallfunc*) > (Py_ssize_t)type->tp_basicsize) {
+        PyErr_Format(PyExc_TypeError,
+                    "tp_basicsize for type '%s' (%d) is too small for vectorcall offset %d",
+                    type->tp_name, type->tp_basicsize, type->tp_vectorcall_offset);
+        return 0;
+    }
+    return 1;
+}
+
 PyObject *
 PyType_FromMetaclass(PyTypeObject *metaclass, PyObject *module,
                      PyType_Spec *spec, PyObject *bases_in)
 {
+    /* Invariant: A non-NULL value in one of these means this function holds
+     * a strong reference or owns allocated memory.
+     * These get decrefed/freed/returned at the end, on both success and error.
+     */
     PyHeapTypeObject *res = NULL;
     PyObject *modname = NULL;
     PyTypeObject *type;
     PyObject *bases = NULL;
+    char *tp_doc = NULL;
+    PyObject *ht_name = NULL;
+    char *_ht_tpname = NULL;
+
     int r;
+    size_t len;
```

Comment:
The number of local variables is getting rather large for this function. Can we try to keep it down by, for example, moving this down to where it is used? (I know there are two different users, but they've got separate scopes.)

<details>

```diff
diff --git a/Objects/typeobject.c b/Objects/typeobject.c
index 343b690470..1d3aaeb89e 100644
--- a/Objects/typeobject.c
+++ b/Objects/typeobject.c
@@ -3452,3 +3452,2 @@ PyType_FromMetaclass(PyTypeObject *metaclass, PyObject *module,
     int r;
-    size_t len;
 
@@ -3516,3 +3515,3 @@ PyType_FromMetaclass(PyTypeObject *metaclass, PyObject *module,
             else {
-                len = strlen(slot->pfunc)+1;
+                size_t len = strlen(slot->pfunc)+1;
                 tp_doc = PyObject_Malloc(len);
@@ -3663,5 +3662,5 @@ PyType_FromMetaclass(PyTypeObject *metaclass, PyObject *module,
             break;
-        case Py_tp_members:
+        case Py_tp_members: {
             /* Move the slots to the heap type itself */
-            len = Py_TYPE(type)->tp_itemsize * nmembers;
+            size_t len = Py_TYPE(type)->tp_itemsize * nmembers;
             memcpy(_PyHeapType_GET_MEMBERS(res), slot->pfunc, len);
@@ -3669,2 +3668,3 @@ PyType_FromMetaclass(PyTypeObject *metaclass, PyObject *module,
             break;
+        }
         default:
```

</details>

Reply:
I formatted it a bit differently to
- avoid the extra `}` being at the same level as the `switch`'s `}`.
- make it more obvious that the `break` is not missing